### DISCUSSION TITLE
How to writeup Liquidsoap file to work with HLS (Icecast/Liquidsoap script provided)

### DISCUSSION THREAD

**Discussion by carrzkiss** [savonet/liquidsoap#3826] [User Support]
Hello, Liquidsoap community.
I have been using **Icecast** with **Liquidsoap** since I started working with media in Linux servers.
I was turned onto **HLS** for being the better option to do what I wanted.
After reading several documents, I found out that I can use a Dynamic URL
and when items are added or subtracted from the url, this will update the .m3u8 file without restarting icecast or HLS.
This is great news, as I get my feed from a database playlist.
The idea is to have the DJ be able to move songs around, add or delete songs, and add jingles, commercials, and so forth from the database. 
So if HLS allows this with liquidsoap, then this is what I will use.

I am currently going through the installation of NGINX and HLS and getting ready to install Liquidsoap.

Question.

Here is an example of my previous LIQ file with Icecast and Liquidsoap.
As you can see, I have two Stations in this example, which ran wonderfully for both myself and my son for quite some time. However, they would have to be restarted for every song list update.

Radio.liq
```
# !/home/iceadmin/.opam/default/bin/liquidsoap
# log dir
set ("log.file.path","/var/log/liquidsoap/basic-radio.log")
set("scheduler.fast_queues",1)
 # Music
#music = playlist("/opt/liquidsoap/playlists/RadioOne.m3u",mode="normal",reload_mode="watch")
Station101  = playlist("https://www.domain.com/Panels/DJ/101.asp",mode="normal")
Station102  = playlist("https://www.domain.com/Panels/DJ/102.asp",mode="normal")

#set("scheduler.fast_queues",1)

# Start building the feed with music
radio101 = Station101
radio102 = Station102

#Station101
output.icecast(%mp3(bitrate=256,samplerate=44100,internal_quality=0,id3v2=true,stereo=true,stereo_mode="stereo"),
name="Radio One",
encoding="UTF-8",
host="192.168.2.203",
port=8000,
password="HACKME",
icy_metadata="true",
description="The Internets Home For Pure Music!",
mount="Radio101",
mksafe(radio101))

#Station102
output.icecast(%mp3(bitrate=256,samplerate=44100,internal_quality=0,id3v2=true,stereo=true,stereo_mode="stereo"),
name="Radio Two",
encoding="UTF-8",
host="192.168.2.203",
port=8000,
password="HACKME",
icy_metadata="true",
description="The place where I come to Rock!",
mount="Radio102",
mksafe(radio102))
```

How would this be written up for HLS with Liquidsoap?
I am looking at Question [#3261](https://github.com/savonet/liquidsoap/discussions/3261)
And I see this.

output.file.hls(playlist="live.m3u8", segment_duration=4.0, segments=10, segments_overhead=5, segment_name=segment_name, "C:/nginx/html/hls", streams, input)

With multiple stations, would the live.m3u8 be changed to reference the name of each individual station? Example: Radio101.m3u8 & Radio102.m3u8 etc..

Thank you to everyone for any assistance you can give on this one.
Wayne

**Comment by gAlleb** [comment]
> However, they would have to be restarted for every song list update.

Hi! As far as I know you can't watch distant playlists however they can be reloaded and you can set time for it with  `playlist(reload=60, "http://playlist")` no matter what `output` you use (icecast or hls).

> and when items are added or subtracted from the url, this will update the .m3u8 file without restarting icecast or HLS.

I don't think this is right. With HLS - M3U8 file is simply text file format containing multimedia playlists, `not your playlist with music`, but playlist of small audio (in our case) files (segments) which are generated by a program so you can stream them via http protocol.

<hr>

To produce HLS streams you have to add to your configuration `output.file.hls`. If you don't need to use Icecast - you can remove `output.icecast` outputs from your configuration or add `output.file.hls` after them like this:

```ruby

radio101 = mksafe(radio101)

# For Station101
# Define 3 streams with different bitrates. You can make 1 or 2 streams (as you wish).

aac_lofi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="96k",
        profile="aac_low"
    )
)
aac_midfi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="128k",
        profile="aac_low"
    )
)
aac_hifi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="192k",
        profile="aac_low"
    )
)

# Put them all together 
hls_streams = [("aac_lofi", aac_lofi), ("aac_midfi", aac_midfi), ("aac_hifi", aac_hifi)]

# Make interesting names for your segments
def hls_segment_name(~position,~extname,stream_name) =
    timestamp = int_of_float(time())
    duration = 4
    "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

output.file.hls(playlist="radio101.m3u8",
    segment_duration=4.0,
    segments=5,
    segments_overhead=2,
    segment_name=hls_segment_name,
    persist_at="C:/nginx/html/hls/persist",
    "C:/nginx/html/hls",
    hls_streams,
    radio101
)
```

<hr>

```ruby

radio102 = mksafe(radio102)

# For Station102
# Define 3 streams with different bitrates
aac_2_lofi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="96k",
        profile="aac_low"
    )
)
aac_2_midfi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="128k",
        profile="aac_low"
    )
)
aac_2_hifi = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="192k",
        profile="aac_low"
    )
)
hls_streams_2 = [("aac_2_lofi", aac_2_lofi), ("aac_2_midfi", aac_2_midfi), ("aac_2_hifi", aac_2_hifi)]

def hls_segment_name_2(~position,~extname,stream_name) =
    timestamp = int_of_float(time())
    duration = 4
    "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

output.file.hls(playlist="radio102.m3u8",
    segment_duration=4.0,
    segments=5,
    segments_overhead=2,
    segment_name=hls_segment_name_2,
    persist_at="C:/nginx/html/hls/persist",
    "C:/nginx/html/hls",
    hls_streams_2,
    radio102
)

```

You can actually read about it in The Book (https://www.liquidsoap.info/book/book.pdf) or https://www.liquidsoap.info/doc-2.2.4/hls_output.html

**Reply by vitoyucepi** [reply]
I would like to point out that if both stations have identical encoders, there is no need to name them differently.

<details>
<summary><code>formats.liq</code></summary>

```ruby
AAC_LOFI = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="96k",
        profile="aac_low"
    )
)
AAC_MIDFI = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="128k",
        profile="aac_low"
    )
)
AAC_HIFI = %ffmpeg(
    format="mpegts",
    %audio(
        codec="aac",
        samplerate=44100,
        channels=2,
        b="192k",
        profile="aac_low"
    )
)
```

</details>

The same applies to naming and outputting.

<details>
<summary><code>hls.liq</code></summary>

```ruby
# Make interesting names for your segments
def hls_segment_name(~position, ~extname, stream_name)
    timestamp = int_of_float(time())
    duration = 4
    "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

def hls_output(streams, source, dir)
    output.file.hls(
        streams,
        source,
        segment_duration=4.0,
        segments=5,
        segments_overhead=2,
        segment_name=hls_segment_name,
        persist_at="#{dir}/state",
        dir,
    )
end
```

</details>

The result will look like this
```ruby
%include "formats.liq"
%include "hls.liq"

hls_streams = [
  ("aac_lofi", AAC_LOFI),
  ("aac_midfi", AAC_MIDFI),
  ("aac_hifi", AAC_HIFI)
]

hls_output(hls_streams, radio101, "C:/nginx/html/hls/101")
hls_output(hls_streams, radio102, "C:/nginx/html/hls/102")
```

**Reply by gAlleb** [reply]
Thank you, Vito! 

Also I remember doing this, especially one `def hls_segment_name` for a few different streams. I don't know why, was it a coincidence or not, but after few hours-days - streams started behaving in a bizzare way, with micro delays and stuttering. But seperating them did the trick. Maybe it's been connected to some `atomic write` issues which have been fixed since.

**Reply by carrzkiss** [reply]
Vito.
How would I load the station?
In the network I am using 
192.168.5.10:8080 (This loads the webpage from Nginx)
I tried adding the station name to the end, but it did not work. ( /radio101 ).

I've installed liquidsoap-daemon.
And have it working, and the liquidsoap user owns the folder.
I ran this
./daemonize-liquidsoap.sh /srv/liquidsoap/101.liq
And it created the `101-run.liq` in the `script` folder.

**Comment by carrzkiss** [comment]
Hey, [Stefan](https://github.com/gAlleb)
I hope you're doing well.
I was told that I could only reload/refresh a distant URL in Icecast if I had many different things added to it for Icecast to reload a URL playlist. It is on another one of my threads here. I will have to find it and tell you what I was told.

I am going to reinstall the Ubuntu with NGINX, HLS, and Liquidsoap.
The one I did last night did not do too well.

Once I get to this point again, I will come in, grab your code, and proceed from there.
**However, I see nothing in the code from which it gets the URL.**
As in my example it is this.
`Station101  = playlist("https://www.domain.com/Panels/DJ/101.asp",mode="normal")`

**Comment by gAlleb** [comment]
Just add my code after yours (I showed you what you can ADD to your current configuration to enable HLS). And if you want to reload playlists (e.g. every 10 minutes) change ` Station101  = playlist("https://www.domain.com/Panels/DJ/101.asp",mode="normal", reload=600)`

**Reply by vitoyucepi** [reply]
Reloading can also be done with `playlist(path, reload_mode="rounds", reload=1)`.

**Comment by carrzkiss** [comment]
Oh, OK.
Thanks, I will work on this later on this evening or tomorrow.
